trafaret/base.py

```diff
@@ -432,30 +433,53 @@ def __repr__(self):
         return "<ToBool>"
 
 
-class Atom(Trafaret):
+class Literal(Trafaret):
     """
-    >>> Atom('atom').check('atom')
+    >>> Literal('atom').check('atom')
     'atom'
-    >>> extract_error(Atom('atom'), 'molecule')
-    "value is not exactly 'atom'"
-    >>> Atom('atom').is_valid('atom')
+    >>> extract_error(Literal('atom'), 'molecule')
+    "value doesn't match any variant"
+    >>> Literal('atom').is_valid('atom')
     True
-    >>> Atom('atom').is_valid('molecule')
+    >>> Literal('atom').is_valid('molecule')
+    False
+    >>> trafaret = Literal('foo', 'bar', 1) >> ignore
+    >>> trafaret
+    <And(<Literal('foo', 'bar', 1)>, <Call(ignore)>)>
+    >>> trafaret.check('foo')
+    >>> trafaret.check(1)
+    >>> extract_error(trafaret, 2)
+    "value doesn't match any variant"
+    >>> trafaret.is_valid(1)
+    True
+    >>> trafaret.is_valid(2)
     False
     """
-    __slots__ = ['value']
+    __slots__ = ['variants']
 
-    def __init__(self, value):
-        self.value = value
+    def __init__(self, *variants):
+        self.variants = variants
 
     def check_value(self, value):
-        if self.value != value:
+        if value not in self.variants:
             self._failure(
-                "value is not exactly '%s'" % self.value,
+                "value doesn't match any variant",
                 value=value,
-                code=codes.IS_NOT_EXACTLY,
+                code=codes.DOES_NOT_MATCH_ANY,
             )
 
+    def __repr__(self):
+        return "<Literal(%s)>" % (", ".join(repr(v) for v in self.variants))
+
+
+class Atom(Literal):
+    """
+    Deprecated: use Literal.
```

Comment:
Why we should deprecate `Atom`? I agree that `Literal` is more genreous, but it is not always that you need this generosity.

Reply:
`Atom("foo")` and `Enum("foo")` do the same thing, there's no need for 2 different classes. That's why there is no Atom type in the typing module. I just thought it would be good to follow the Python zen:

> There should be one-- and preferably only one --obvious way to do it.

It also keeps things consistent with the typing module, which means anyone familiar with that would already know what to expect with Trafaret (given that they are quite similar elsewhere).

Reply:
I do not see why
1. everything should be consistent with typing module – it is not typing, not about typing, and was there long before typing.
2. if Atom will be there you are free to not use it. I would.

Reply:
To me, this library seems like runtime type checking, with some extra features to handle transformation etc.
If I have the time, I'd like to add support for using type annotations directly in trafaret for this, e.g. defining a trafaret dict something like this (based on the typing documentation, final interface could look a lot different):
```
class Foo(TypedDict):
    type: Literal["spam", "eggs"]
    bar: Annotated[int, ValueRange(-10, 5)]

res = t.Dict(Foo).check(a)  # return type is Foo
res["bar"] + "spam"  # Mypy would give a type error here.
```

This has the benefit of defining the details needed for trafaret at runtime, but also defining the return type of the trafaret check for static typing. While it doesn't require changing the trafaret objects, it would be more consistent and obvious as to how the library works (while avoiding the confusion with Enum).

As for the second point, that doesn't really explain why Atom exists. I'm just saying it seems pointless to have another object that adds no functionality. It's not obvious to me why I would use Atom instead of Enum.

Reply:
The annotations feature is something I'm likely to work on beginning of next year at work.

My initial ideas around the interface is to support something like:
```
class Something(TypedDict):
    type: Literal["spam", "eggs"]
    bar: Annotated[int, t.Args(gte=0, lte=10)]
    value: Annotated[str, t.Args(allow_blank=True)]

res = t.Dict(Something).check(...)
```
Which would basically just map the types to trafaret objects (e.g. str => String, int => Int). `Args()` would then be used to pass any arguments to the trafaret object.

I think this could also be extended checking function arguments as well, something like:
```
@t.check
def some_thing(some_id: Annotated[int, t.Args(gte=0)], value: Annotated[str, t.Arg[min_length=5)]):
    ...

some_thing(-5, "")  # Raises a trafaret error.
```